go/cmd/server/logging/logging.go

```diff
@@ -0,0 +1,306 @@
+package logging
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/apache/arrow/go/v8/arrow"
+	"github.com/apache/arrow/go/v8/arrow/array"
+	"github.com/apache/arrow/go/v8/arrow/memory"
+	"github.com/feast-dev/feast/go/internal/feast"
+	"github.com/feast-dev/feast/go/internal/feast/model"
+	"github.com/feast-dev/feast/go/protos/feast/serving"
+	"github.com/feast-dev/feast/go/protos/feast/types"
+	gotypes "github.com/feast-dev/feast/go/types"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type Log struct {
+	// Example: val{int64_val: 5017}, val{int64_val: 1003}
+	EntityValue []*types.Value
+
+	FeatureValues   []*types.Value
+	FeatureStatuses []serving.FieldStatus
+	EventTimestamps []*timestamppb.Timestamp
+	RequestContext  map[string]*types.RepeatedValue
+}
+
+type MemoryBuffer struct {
+	featureService *model.FeatureService
+	logs           []*Log
+}
+
+type LoggingService struct {
+	memoryBuffer      *MemoryBuffer
+	logChannel        chan *Log
+	fs                *feast.FeatureStore
+	offlineLogStorage OfflineLogStorage
+}
+
+func NewLoggingService(fs *feast.FeatureStore, logChannelCapacity int, featureServiceName string, enableLogging bool) (*LoggingService, error) {
+	var featureService *model.FeatureService = nil
+	var err error
+	if enableLogging {
+		featureService, err = fs.GetFeatureService(featureServiceName)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	loggingService := &LoggingService{
+		logChannel: make(chan *Log, logChannelCapacity),
+		memoryBuffer: &MemoryBuffer{
+			logs:           make([]*Log, 0),
+			featureService: featureService,
+		},
+		fs: fs,
+	}
+
+	if !enableLogging || fs == nil {
+		loggingService.offlineLogStorage = nil
+	} else {
+		offlineLogStorage, err := NewOfflineStore(fs.GetRepoConfig())
+		loggingService.offlineLogStorage = offlineLogStorage
+
+		if err != nil {
+			return nil, err
+		}
+		// Start goroutine to process logs
+		go loggingService.processLogs()
+	}
+	return loggingService, nil
+}
+
+func (s *LoggingService) EmitLog(l *Log) error {
+	select {
+	case s.logChannel <- l:
+		return nil
+	case <-time.After(20 * time.Millisecond):
```

Comment:
nitpick: would be good to parametrize this timeout

Reply:
Done.